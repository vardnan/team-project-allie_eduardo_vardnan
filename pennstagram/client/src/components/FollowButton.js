import React, { useState } from 'react';
import PropTypes from 'prop-types';
// import { UserContext } from '../contexts/user-context';
import io from 'socket.io-client';
import { updateUserFollowing } from '../api/updateUserFollowing';

function FollowButton({ viewingUsername }) {
  // const { currentUser, setCurrentUser } = useContext(UserContext);
  const currentUser = JSON.parse(localStorage.getItem('user'));
  // const socket = io('http://localhost:8080');
  const socket = io('');

  const [isFollowing, setIsFollowing] = useState(
    currentUser.following
      ? currentUser.following.includes(viewingUsername)
      : false,
  );

  const toggleFollow = async () => {
    try {
      let updatedFollowingList = [];

      if (isFollowing) {
        // If already following, unfollow
        updatedFollowingList = currentUser.following.filter(
          (username) => username !== viewingUsername,
        );
        // console.log('this is the following list now');
        // console.log(currentUser.username);
        // console.log(viewingUsername);
        // console.log(updatedFollowingList);
      } else {
        // If not following, follow
        updatedFollowingList = currentUser.following
          ? [...currentUser.following, viewingUsername] : [viewingUsername];
      }

      // Update the user's following list using the new function
      // await updateUserFollowing(currentUser.username, updatedFollowingList);

      // Update the user's following list using the new function
      // eslint-disable-next-line
      // console.log(currentUser.username);
      // eslint-disable-next-line
      // console.log(viewingUsername);
      try {
        await updateUserFollowing(currentUser.username, viewingUsername, updatedFollowingList);
      } catch (error) {
        // eslint-disable-next-line
        console.log(error);
      }

      // Emit a follower update event
      socket.emit('followerUpdate', {
        username: viewingUsername,
      });

      // Toggle the isFollowing state
      setIsFollowing(!isFollowing);
      // setCurrentUser({
      //   ...currentUser,
      //   following: updatedFollowingList,
      // });
      currentUser.following = updatedFollowingList;
      localStorage.setItem('user', JSON.stringify(currentUser));
    } catch (error) {
      // eslint-disable-next-line
      console.error('Error toggling follow:', error);
    }
  };

  return (
    <div>
      <button type="button" onClick={toggleFollow}>
        {' '}
        {/* Add type="button" */}
        {isFollowing ? 'Unfollow' : 'Follow'}
      </button>
    </div>
  );
}

FollowButton.propTypes = {
  viewingUsername: PropTypes.string.isRequired,
};

export default FollowButton;
