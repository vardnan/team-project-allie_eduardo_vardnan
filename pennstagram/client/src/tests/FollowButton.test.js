import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import FollowButton from '../components/FollowButton';
import UserContext from '../contexts/user-context';

// Mock functions
const mockFollowUser = jest.fn();
const mockUnfollowUser = jest.fn();

// Mock Context
const mockContext = {
  currentUser: { following: ['Bob'] },
  followUser: mockFollowUser,
  unfollowUser: mockUnfollowUser,
};

// Testing rendering
test('Testing that the FollowButton is rendering correctly when already following', () => {
  const { getByText } = render(
    <UserContext.Provider value={mockContext}>
      <FollowButton viewingUsername="Bob" />
    </UserContext.Provider>
  );
  expect(getByText('Unfollow')).toBeInTheDocument();
});

test('Testing that the FollowButton is rendering correctly when not following', () => {
  const { getByText } = render(
    <UserContext.Provider value={mockContext}>
      <FollowButton viewingUsername="Alice" />
    </UserContext.Provider>
  );
  expect(getByText('Follow')).toBeInTheDocument();
});

// Snapshot testing
test('Testing that the FollowButton matches snapshot', () => {
  const component = renderer.create(
    <UserContext.Provider value={mockContext}>
      <FollowButton viewingUsername="Bob" />
    </UserContext.Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// Testing events
test('Testing the toggleFollow function when initially following', () => {
  const { getByText } = render(
    <UserContext.Provider value={mockContext}>
      <FollowButton viewingUsername="Bob" />
    </UserContext.Provider>
  );
  fireEvent.click(getByText('Unfollow'));
  expect(mockUnfollowUser).toHaveBeenCalledWith('Bob');
  expect(getByText('Follow')).toBeInTheDocument();
});

test('Testing the toggleFollow function when initially not following', () => {
  const { getByText } = render(
    <UserContext.Provider value={mockContext}>
      <FollowButton viewingUsername="Alice" />
    </UserContext.Provider>
  );
  fireEvent.click(getByText('Follow'));
  expect(mockFollowUser).toHaveBeenCalledWith('Alice');
  expect(getByText('Unfollow')).toBeInTheDocument();
});
