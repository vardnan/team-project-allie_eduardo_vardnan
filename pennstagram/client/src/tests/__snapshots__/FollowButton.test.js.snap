// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing that the FollowButton matches snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Unfollow
  </button>
</div>
`;
